java/ql/src/semmle/code/java/frameworks/apache/Collections.qll

```diff
@@ -0,0 +1,212 @@
+    this.hasName("isEmpty")
+  }
+}
+
+/**
+ * The method `isNotEmpty` in either `org.apache.commons.collections.CollectionUtils`
+ * or `org.apache.commons.collections4.CollectionUtils`.
+ */
+class MethodApacheCollectionsIsNotEmpty extends Method {
+  MethodApacheCollectionsIsNotEmpty() {
+    this.getDeclaringType()
+        .hasQualifiedName(["org.apache.commons.collections", "org.apache.commons.collections4"],
+          "CollectionUtils") and
+    this.hasName("isNotEmpty")
+  }
+}
+
+/**
+ * Value-propagating models for classes in the package `org.apache.commons.collections4`.
+ */
+private class ApacheCollectionsModel extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "org.apache.commons.collections4;KeyValue;true;getKey;;;MapKey of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4;KeyValue;true;getValue;;;MapValue of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections;KeyValue;true;getKey;;;MapKey of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections;KeyValue;true;getValue;;;MapValue of Argument[-1];ReturnValue;value"
+      ]
+  }
+}
+
+/**
+ * Value-propagating models for classes in the package `org.apache.commons.collections4.keyvalue`.
+ */
+private class ApacheKeyValueModel extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "org.apache.commons.collections4.keyvalue;AbstractKeyValue;true;AbstractKeyValue;;;Argument[0];MapKey of Argument[-1];value",
+        "org.apache.commons.collections4.keyvalue;AbstractKeyValue;true;AbstractKeyValue;;;Argument[1];MapValue of Argument[-1];value",
+        "org.apache.commons.collections4.keyvalue;AbstractKeyValue;true;setKey;;;MapValue of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4.keyvalue;AbstractKeyValue;true;setKey;;;Argument[0];MapKey of Argument[-1];value",
+        "org.apache.commons.collections4.keyvalue;AbstractKeyValue;true;setValue;;;MapValue of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4.keyvalue;AbstractKeyValue;true;setValue;;;Argument[0];MapValue of Argument[-1];value",
+        "org.apache.commons.collections4.keyvalue;AbstractKeyValue;true;toString;;;MapKey of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4.keyvalue;AbstractKeyValue;true;toString;;;MapValue of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4.keyvalue;AbstractMapEntry;true;AbstractMapEntry;;;Argument[0];MapKey of Argument[-1];value",
+        "org.apache.commons.collections4.keyvalue;AbstractMapEntry;true;AbstractMapEntry;;;Argument[1];MapValue of Argument[-1];value",
+        "org.apache.commons.collections4.keyvalue;AbstractMapEntryDecorator;true;AbstractMapEntryDecorator;;;Argument[0];Argument[-1];value",
+        "org.apache.commons.collections4.keyvalue;AbstractMapEntryDecorator;true;getMapEntry;;;Argument[-1];ReturnValue;value",
```

Comment:
Search through all of these csv rows for `;[^ ]*;[^ ]*;value"` - I predict that all rows matching this regex will be wrong (because without `Foo of Bar` somewhere in an input or output spec then it means that the implementation is the identity function `foo(x) { return x; }`, which I doubt is the case).

Reply:
Most of them are wrong, but some are correct. They are `MapUtils.emptyIfNull`, `MapUtils.getMap`, `MapUtils.getObject` and `MapUtils.getString`. In all cases they are the identity function in some situation, e.g. you pass a default value to be returned if the key doesn't exist in the map.